eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Converts an array  to a list data structure
+ * @param {*} inputArray 
+ * @param {*} list 
+ * @returns 
+ */
+function arrayToList([...inputArray], list = { value: null, rest: null }) { 
+    if (!Array.isArray(inputArray)) { //this needs work
+        throw "input must be array";
+    }
+
+    if (inputArray.length === 0) {
+        return list.rest;
+    }
+
+    list.value = inputArray.pop();
+    list.rest = structuredClone(list);
+
+    return arrayToList(inputArray, list);
+}
+
+/**
+ * Converts a list data structure to an array
+ * @param {*} inputList 
+ * @param {*} arrayAccumulator 
+ * @returns 
+ */
+function listToArray(inputList, arrayAccumulator = []) {
+    if (!Object.keys(inputList).includes('value') || !Object.keys(inputList).includes('rest')) {
+        throw "input must be list data structure";
+    }
+    
+    arrayAccumulator.push(inputList.value);
+
+    if (inputList.rest) {
+        listToArray(inputList.rest,arrayAccumulator)
```

Comment:
I think it would be good practice to also do this using looping. The solutions represent the same mathematical idea, but being able to convert between them in your head is valuable. It's also almost always the case that the looping solution is faster, but the recursive solution is often enough easier to implement and maintain that the small difference is worth it.